drivers/mdio/mdio_xilinx_axienet.c

```diff
@@ -0,0 +1,200 @@
+	mdio_bus_wait_ready(config);
+	data->bus_enabled = true;
+}
+
+static void mdio_xilinx_axienet_bus_enable(const struct device *dev)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *data = dev->data;
+	uint16_t clock_divider, clock_divider_full;
+
+	if (!config->clock_frequency_hz) {
+		LOG_ERR("No clock frequency specified for ethernet device!");
+		return;
+	}
+
+	/* this might result in a MDIO frequency that is a bit lower than the max frequency */
+	clock_divider_full = (config->clock_frequency_hz / (MDIO_TARGET_FREQUENCY_HZ)) * 2;
+	clock_divider = clock_divider_full & XILINX_AXIENET_MDIO_SETUP_REG_MDIO_CLOCK_DIVIDER_MASK;
+
+	if (clock_divider != clock_divider_full) {
+		LOG_ERR("Clock divider overflow!");
+	}
+
+	data->clock_divider = clock_divider;
+
+	LOG_INF("Enable MDIO Bus assuming ethernet clock frequency %u divider %u!",
+		config->clock_frequency_hz, clock_divider);
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_SETUP_REG_OFFSET,
+					   clock_divider);
+
+	enable_mdio_bus(config, data);
+
+	LOG_INF("MDIO ready!");
+}
+
+static int mdio_xilinx_axienet_read(const struct device *dev, uint8_t prtad, uint8_t devad,
+				    uint16_t *data)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *dev_data = dev->data;
+
+	enable_mdio_bus(config, dev_data);
+	xilinx_axienet_mdio_write_register(
+		config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET,
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE |
+			(prtad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR) |
+			(devad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR) |
+			XILINX_AXIENET_MDIO_CONTROL_REG_MASK_READ);
+
+	mdio_bus_wait_ready(config);
+
+	*data = (uint16_t)(xilinx_axienet_read_mdio_register(
+				   config, XILINX_AXIENET_MDIO_READ_DATA_REG_OFFSET) &
+			   XILINX_AXIENET_MDIO_READ_DATA_REG_DATA_MASK);
+
+	LOG_DBG("Read %" PRIu16 " from MDIO!", *data);
+
+	return 0;
+}
+
+static int mdio_xilinx_axienet_write(const struct device *dev, uint8_t prtad, uint8_t devad,
+				     uint16_t data)
+{
+	const struct mdio_xilinx_axienet_config *config = dev->config;
+	struct mdio_xilinx_axienet_data *dev_data = dev->data;
+
+	xilinx_axienet_mdio_write_register(config, XILINX_AXIENET_MDIO_WRITE_DATA_REG_OFFSET, data);
+
+	enable_mdio_bus(config, dev_data);
+	xilinx_axienet_mdio_write_register(
+		config, XILINX_AXIENET_MDIO_CONTROL_REG_OFFSET,
+		XILINX_AXIENET_MDIO_CONTROL_REG_MASK_INITIATE |
+			(prtad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_PHYADDR) |
+			(devad << XILINX_AXIENET_MDIO_CONTROL_REG_SHIFT_REGADDR) |
+			XILINX_AXIENET_MDIO_CONTROL_REG_MASK_WRITE);
+
+	mdio_bus_wait_ready(config);
+
+	return 0;
+}
+
+static const struct mdio_driver_api mdio_xilinx_axienet_api = {
+	.bus_disable = mdio_xilinx_axienet_bus_disable,
+	.bus_enable = mdio_xilinx_axienet_bus_enable,
+	.read = mdio_xilinx_axienet_read,
+	.write = mdio_xilinx_axienet_write};
```

Comment:
nit please put the closing bracket on a new line

Reply:
Done